File: moqt/track_reader_test.go
package moqt

import (
	"bytes"
	"context"
	"sync/atomic"
	"testing"
	"time"

	"github.com/qumo-dev/gomoqt/moqt/internal/message"
	"github.com/qumo-dev/gomoqt/transport"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func newTestTrackReader(tb testing.TB) (*TrackReader, *FakeQUICStream) {
	tb.Helper()
	mockStream := &FakeQUICStream{}

	substr := newTestSendSubscribeStreamFromStream(mockStream, &SubscribeConfig{})
	receiver := newTrackReader("/test", "video", substr, func() {})
	return receiver, mockStream
}

// blockingReceiveStream is a FakeQUICReceiveStream whose Read blocks until
// CancelRead is invoked, then returns a *transport.StreamError — modeling the
// quic-go/webtransport guarantee that CancelRead unblocks an in-flight Read.
func blockingReceiveStream(canceled *atomic.Bool) *FakeQUICReceiveStream {
	release := make(chan struct{})
	return &FakeQUICReceiveStream{
		ReadFunc: func(p []byte) (int, error) {
			<-release
			return 0, &transport.StreamError{ErrorCode: transport.StreamErrorCode(SubscribeCanceledErrorCode)}
		},
		CancelReadFunc: func(code transport.StreamErrorCode) {
			canceled.Store(true)
			select {
			case <-release:
			default:
				close(release)
			}
		},
	}
}

// TestTrackReader_Close_CancelsAcceptedGroup is a regression test for the dead
// `dequeued` tracking: an already-accepted GroupReader was tracked nowhere, so
// Close could not cancel it and a consumer parked in ReadFrame blocked forever.
// It must now be canceled via the groupManager.
func TestTrackReader_Close_CancelsAcceptedGroup(t *testing.T) {
	receiver, _ := newTestTrackReader(t)

	var canceled atomic.Bool
	groupStream := blockingReceiveStream(&canceled)

	receiver.enqueueGroup(GroupSequence(1), groupStream)

	gr, err := receiver.AcceptGroup(context.Background())
	require.NoError(t, err)
	require.NotNil(t, gr)

	readDone := make(chan error, 1)
	go func() {
		readDone <- gr.ReadFrame(NewFrame(0))
	}()

	require.NoError(t, receiver.Close())

	assert.True(t, canceled.Load(), "accepted group stream should be canceled on Close")
	select {
	case err := <-readDone:
		assert.Error(t, err, "parked ReadFrame should unblock after Close")
	case <-time.After(2 * time.Second):
		t.Fatal("ReadFrame did not unblock after Close (deadlock regression)")
	}
}

func TestNewTrackReader(t *testing.T) {
	mockStream := &FakeQUICStream{}
	substr := newSendSubscribeStream(SubscribeID(1), mockStream, &SubscribeConfig{})
	receiver := newTrackReader("/test", "video", substr, func() {})

	assert.NotNil(t, receiver, "newTrackReader should not return nil")
	assert.Equal(t, BroadcastPath("/test"), receiver.BroadcastPath)
	assert.Equal(t, TrackName("video"), receiver.TrackName)
	// Verify info propagation
	assert.Equal(t, PublishInfo{}, substr.ReadInfo(), "sendSubscribeStream should return the Info passed at construction")
	assert.NotNil(t, receiver.queueing, "queue should be initialized")
	assert.NotNil(t, receiver.queuedCh, "queuedCh should be initialized")
	assert.NotNil(t, receiver.groupManager, "groupManager should be initialized")
}

func TestTrackReader_AcceptGroup(t *testing.T) {
	receiver, _ := newTestTrackReader(t)

	// Test with a timeout to ensure we don't block forever when no groups are available
	testCtx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
	defer cancel()

	_, err := receiver.AcceptGroup(testCtx)
	assert.Error(t, err, "expected timeout error when no groups are available")
	assert.Equal(t, context.DeadlineExceeded, err, "expected deadline exceeded error")
}

func TestTrackReader_ContextCancellation(t *testing.T) {
	ctx, cancel := context.WithCancel(context.Background())
	mockStream := &FakeQUICStream{
		ParentCtx: ctx,
	}
	substr := newTestSendSubscribeStreamFromStream(mockStream, &SubscribeConfig{})
	receiver := newTrackReader("/test", "video", substr, func() {})

	// Cancel the context
	cancel()

	// Test that AcceptGroup returns context error when context is cancelled
	testCtx, testCancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
	defer testCancel()

	_, err := receiver.AcceptGroup(testCtx)
	assert.Error(t, err, "expected error when context is cancelled")
	// Should return context.Canceled or DeadlineExceeded
	assert.True(t, err == context.Canceled || err == context.DeadlineExceeded, "expected context error")
}

func TestTrackReader_Context_FollowsStreamLifecycle(t *testing.T) {
	_, cancelSetup := context.WithCancel(context.Background())
	defer cancelSetup()

	mockStream := &FakeQUICStream{}

	substr := newTestSendSubscribeStreamFromStream(mockStream, &SubscribeConfig{})
	receiver := newTrackReader("/test", "video", substr, func() {})

	// Cancel setup context; TrackReader context should remain alive while stream is alive.
	cancelSetup()
	select {
	case <-receiver.Context().Done():
		t.Fatal("track reader context should not be canceled by request setup context")
	case <-time.After(20 * time.Millisecond):
		// expected
	}

	// Close stream; TrackReader context should be canceled.
	require.NoError(t, mockStream.Close())

	select {
	case <-receiver.Context().Done():
		// expected
	case <-time.After(100 * time.Millisecond):
		t.Fatal("track reader context should be canceled when stream is closed")
	}
}

func TestTrackReader_EnqueueGroup(t *testing.T) {
	receiver, _ := newTestTrackReader(t)

	// Mock receive stream
	mockReceiveStream := &FakeQUICReceiveStream{}

	// Enqueue a group
	receiver.enqueueGroup(GroupSequence(1), mockReceiveStream)

	// Test that we can accept the enqueued group
	testCtx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
	defer cancel()

	group, err := receiver.AcceptGroup(testCtx)
	assert.NoError(t, err, "should be able to accept enqueued group")
	assert.NotNil(t, group, "accepted group should not be nil")

}

func TestTrackReader_AcceptGroup_RealImplementation(t *testing.T) {
	receiver, _ := newTestTrackReader(t)

	// Test with a timeout to ensure we don't block forever
	testCtx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
	defer cancel()

	_, err := receiver.AcceptGroup(testCtx)
	assert.Error(t, err, "expected timeout error when no groups are available")
	assert.Equal(t, context.DeadlineExceeded, err, "expected deadline exceeded error")
}

func TestTrackReader_Close(t *testing.T) {
	receiver, _ := newTestTrackReader(t)

	err := receiver.Close()
	assert.NoError(t, err)

	// Close again should not error
	err = receiver.Close()
	assert.NoError(t, err)
}

func TestTrackReader_Update(t *testing.T) {
	receiver, _ := newTestTrackReader(t)

	newTrackConfig := SubscribeConfig{}

	_ = receiver.Update(&newTrackConfig)

	// Verify update
	assert.Equal(t, &SubscribeConfig{}, receiver.TrackConfig())
}

func TestTrackReader_AcceptDrop(t *testing.T) {
	var buf bytes.Buffer
	_, _ = buf.Write([]byte{byte(message.MessageTypeSubscribeDrop)})
	require.NoError(t, (message.SubscribeDropMessage{
		StartGroup: 11,
		EndGroup:   21,
		ErrorCode:  3,
	}).Encode(&buf))

	mockStream := &FakeQUICStream{
		ReadFunc: buf.Read,
	}

	substr := newSendSubscribeStream(SubscribeID(1), mockStream, &SubscribeConfig{})
	receiver := newTrackReader("/test", "video", substr, func() {})

	go substr.readSubscribeResponses()

	ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
	defer cancel()

	drop, err := receiver.acceptDrop(ctx)
	require.NoError(t, err)
	assert.Equal(t, SubscribeDrop{
		StartGroup: 10,
		EndGroup:   20,
		ErrorCode:  3,
	}, drop)
}

func TestTrackReader_CloseWithError(t *testing.T) {
	receiver, _ := newTestTrackReader(t)

	receiver.CloseWithError(SubscribeErrorCodeInternal)
}

func TestGroupReader_CancelRead_RemovesFromManager(t *testing.T) {
	receiver, _ := newTestTrackReader(t)

	recvStream := &FakeQUICReceiveStream{}
	group := newGroupReader(GroupSequence(1), recvStream, receiver.groupManager)

	assert.Len(t, receiver.groupManager.activeGroups, 1)
	assert.Contains(t, receiver.groupManager.activeGroups, group)

	group.CancelRead(InternalGroupErrorCode)
	assert.Len(t, receiver.groupManager.activeGroups, 0)
	assert.NotContains(t, receiver.groupManager.activeGroups, group)
}

func TestTrackReader_Drops(t *testing.T) {
	var buf bytes.Buffer

	// Write a SUBSCRIBE_DROP response (readSubscribeResponses returns after one drop)
	_, _ = buf.Write([]byte{byte(message.MessageTypeSubscribeDrop)})
	require.NoError(t, (message.SubscribeDropMessage{
		StartGroup: 11, // wire value 11 → groupSequenceFromWire → GroupSequence(10)
		EndGroup:   21, // wire value 21 → groupSequenceFromWire → GroupSequence(20)
		ErrorCode:  3,
	}).Encode(&buf))

	mockStream := &FakeQUICStream{
		ReadFunc: buf.Read,
	}

	substr := newSendSubscribeStream(SubscribeID(1), mockStream, &SubscribeConfig{})
	receiver := newTrackReader("/test", "video", substr, func() {})

	go substr.readSubscribeResponses()

	ctx, cancel := context.WithTimeout(context.Background(), 200*time.Millisecond)
	defer cancel()

	var drops []SubscribeDrop
	for drop := range receiver.Drops(ctx) {
		drops = append(drops, drop)
	}

	require.Len(t, drops, 1)
	assert.Equal(t, GroupSequence(10), drops[0].StartGroup)
	assert.Equal(t, GroupSequence(20), drops[0].EndGroup)
	assert.Equal(t, SubscribeErrorCode(3), drops[0].ErrorCode)
}

func TestTrackReader_Drops_ContextCanceled(t *testing.T) {
	mockStream := &FakeQUICStream{}

	substr := newSendSubscribeStream(SubscribeID(1), mockStream, &SubscribeConfig{})
	receiver := newTrackReader("/test", "video", substr, func() {})

	go substr.readSubscribeResponses()

	ctx, cancel := context.WithCancel(context.Background())
	cancel() // cancel immediately

	var drops []SubscribeDrop
	for drop := range receiver.Drops(ctx) {
		drops = append(drops, drop)
	}

	assert.Empty(t, drops)
}

func TestTrackReader_Update_NilConfig(t *testing.T) {
	receiver, _ := newTestTrackReader(t)

	err := receiver.Update(nil)
	assert.Error(t, err)
}

func TestTrackReader_SubscribeID(t *testing.T) {
	mockStream := &FakeQUICStream{}

	substr := newSendSubscribeStream(SubscribeID(42), mockStream, &SubscribeConfig{})
	receiver := newTrackReader("/test", "video", substr, func() {})

	assert.Equal(t, SubscribeID(42), receiver.SubscribeID())
}

File: moq-web/src/track_reader_test.ts
import { assertEquals, assertExists } from "@std/assert";
import { TrackReader } from "./track_reader.ts";
import { Queue } from "./internal/queue.ts";
import { background, withCancelCause } from "@okdaichi/golikejs/context";
import { MockReceiveStream, MockStream } from "./mock_stream_test.ts";
import { SendSubscribeStream } from "./subscribe_stream.ts";
import { GroupMessage, SubscribeMessage, SubscribeOkMessage } from "./internal/message/mod.ts";

Deno.test("TrackReader", async (t) => {
	await t.step(
		"TrackReader.acceptGroup returns a group or error on empty dequeue",
		async () => {
			const [ctx] = withCancelCause(background());
			const stream = new MockStream({});
			const subscribe = new SubscribeMessage({
				subscribeId: 33,
				broadcastPath: "/test",
				trackName: "name",
				subscriberPriority: 1,
			});
			const ok = new SubscribeOkMessage({});
			const sss = new SendSubscribeStream(ctx, stream, subscribe, ok);
			// Use empty queue which causes acceptGroup to return error due to resolved signal
			const queue = new Queue<[any, any]>();
			const onClose = () => {};
			const tr = new TrackReader("/test", "name", sss, queue, onClose);
			const [grp, err] = await tr.acceptGroup(Promise.resolve());
			assertEquals(grp, undefined);
			assertEquals(err instanceof Error, true);
		},
	);

	await t.step(
		"TrackReader.update proxies to subscribeStream update and readInfo returns info",
		async () => {
			const [ctx] = withCancelCause(background());
			const stream = new MockStream({});
			const subscribe = new SubscribeMessage({
				subscribeId: 21,
				broadcastPath: "/test",
				trackName: "name",
				subscriberPriority: 0,
			});
			const ok = new SubscribeOkMessage({});
			const sss = new SendSubscribeStream(ctx, stream, subscribe, ok);
			const queue = new Queue<[any, any]>();
			const onClose = () => {};
			const tr = new TrackReader("/test", "name", sss, queue, onClose);
			const err = await tr.update({
				priority: 5,
				ordered: false,
				maxLatency: 0,
				startGroup: 0,
				endGroup: 0,
			});
			assertEquals(err, undefined);
			assertEquals(tr.readInfo(), sss.info);
		},
	);

	await t.step(
		"TrackReader.acceptGroup returns error when context is cancelled",
		async () => {
			const [ctx, cancel] = withCancelCause(background());
			const stream = new MockStream({});
			const subscribe = new SubscribeMessage({
				subscribeId: 44,
				broadcastPath: "/test",
				trackName: "name",
				subscriberPriority: 1,
			});
			const ok = new SubscribeOkMessage({});
			const sss = new SendSubscribeStream(ctx, stream, subscribe, ok);
			const queue = new Queue<[any, any]>();
			const onClose = () => {};
			const tr = new TrackReader("/test", "name", sss, queue, onClose);

			// Cancel context before calling acceptGroup
			cancel(new Error("context cancelled"));
			await new Promise((r) => setTimeout(r, 0));

			const [grp, err] = await tr.acceptGroup(new Promise(() => {}));
			assertEquals(grp, undefined);
			assertEquals(err instanceof Error, true);
		},
	);

	await t.step(
		"TrackReader.acceptGroup returns GroupReader when group is available",
		async () => {
			const [ctx] = withCancelCause(background());
			const stream = new MockStream({});
			const subscribe = new SubscribeMessage({
				subscribeId: 55,
				broadcastPath: "/test",
				trackName: "name",
				subscriberPriority: 1,
			});
			const ok = new SubscribeOkMessage({});
			const sss = new SendSubscribeStream(ctx, stream, subscribe, ok);
			const queue = new Queue<[any, any]>();
			const onClose = () => {};
			const tr = new TrackReader("/test", "name", sss, queue, onClose);

			// Add a group to the queue
			const mockReceiveStream = new MockReceiveStream({});
			const groupMsg = new GroupMessage({
				subscribeId: 55,
				sequence: 1,
			});
			queue.enqueue([mockReceiveStream, groupMsg]);

			const [grp, err] = await tr.acceptGroup(new Promise(() => {}));
			assertEquals(err, undefined);
			assertExists(grp);
		},
	);

	await t.step("TrackReader.closeWithError calls onCloseFunc", async () => {
		const [ctx] = withCancelCause(background());
		const stream = new MockStream({});
		const subscribe = new SubscribeMessage({
			subscribeId: 66,
			broadcastPath: "/test",
			trackName: "name",
			subscriberPriority: 1,
		});
		const ok = new SubscribeOkMessage({});
		const sss = new SendSubscribeStream(ctx, stream, subscribe, ok);
		const queue = new Queue<[any, any]>();
		let closeWithErrorCalled = false;
		const onClose = () => {
			closeWithErrorCalled = true;
		};
		const tr = new TrackReader("/test", "name", sss, queue, onClose);

		await tr.closeWithError(1);
		assertEquals(closeWithErrorCalled, true);
	});

	await t.step("TrackReader.trackConfig returns correct config", () => {
		const [ctx] = withCancelCause(background());
		const stream = new MockStream({});
		const subscribe = new SubscribeMessage({
			subscribeId: 77,
			broadcastPath: "/test",
			trackName: "name",
			subscriberPriority: 5,
		});
		const ok = new SubscribeOkMessage({});
		const sss = new SendSubscribeStream(ctx, stream, subscribe, ok);
		const queue = new Queue<[any, any]>();
		const tr = new TrackReader("/test", "name", sss, queue, () => {});

		const config = tr.trackConfig;
		assertEquals(config.priority, 5);
	});

	await t.step("TrackReader.context returns subscribe stream context", () => {
		const [ctx] = withCancelCause(background());
		const stream = new MockStream({});
		const subscribe = new SubscribeMessage({
			subscribeId: 88,
			broadcastPath: "/test",
			trackName: "name",
			subscriberPriority: 1,
		});
		const ok = new SubscribeOkMessage({});
		const sss = new SendSubscribeStream(ctx, stream, subscribe, ok);
		const queue = new Queue<[any, any]>();
		const tr = new TrackReader("/test", "name", sss, queue, () => {});

		assertExists(tr.context);
		assertEquals(tr.context.err(), undefined);
	});

	await t.step("TrackReader.close calls onCloseFunc", async () => {
		const [ctx] = withCancelCause(background());
		const stream = new MockStream({});
		const subscribe = new SubscribeMessage({
			subscribeId: 90,
			broadcastPath: "/test",
			trackName: "name",
			subscriberPriority: 1,
		});
		const ok = new SubscribeOkMessage({});
		const sss = new SendSubscribeStream(ctx, stream, subscribe, ok);
		const queue = new Queue<[any, any]>();
		let closeCalled = false;
		const tr = new TrackReader("/test", "name", sss, queue, () => {
			closeCalled = true;
		});

		await tr.close();
		assertEquals(closeCalled, true);
	});

	await t.step("TrackReader.subscribeId returns subscribe stream ID", () => {
		const [ctx] = withCancelCause(background());
		const stream = new MockStream({});
		const subscribe = new SubscribeMessage({
			subscribeId: 42,
			broadcastPath: "/test",
			trackName: "name",
			subscriberPriority: 1,
		});
		const ok = new SubscribeOkMessage({});
		const sss = new SendSubscribeStream(ctx, stream, subscribe, ok);
		const queue = new Queue<[any, any]>();
		const tr = new TrackReader("/test", "name", sss, queue, () => {});

		assertEquals(tr.subscribeId, 42);
	});

	await t.step(
		"TrackReader.drops yields drops and exits on context cancel",
		async () => {
			const [ctx, cancel] = withCancelCause(background());
			const stream = new MockStream({});
			const subscribe = new SubscribeMessage({
				subscribeId: 100,
				broadcastPath: "/test",
				trackName: "name",
				subscriberPriority: 1,
			});
			const ok = new SubscribeOkMessage({});
			const sss = new SendSubscribeStream(ctx, stream, subscribe, ok);
			const queue = new Queue<[any, any]>();
			const tr = new TrackReader("/test", "name", sss, queue, () => {});

			// Append a drop before iterating
			sss.appendDrop({ startGroup: 1, endGroup: 5, errorCode: 0x03 });

			const collected: Array<{ startGroup: number; endGroup: number; errorCode: number }> = [];

			// Cancel context after first yield to exit the generator
			const signal = new Promise<void>(() => {});
			for await (const drop of tr.drops(signal)) {
				collected.push(drop);
				// Cancel context to stop the generator
				cancel(new Error("done"));
				break;
			}

			assertEquals(collected.length, 1);
			assertEquals(collected[0]!.startGroup, 1);
			assertEquals(collected[0]!.endGroup, 5);
			assertEquals(collected[0]!.errorCode, 0x03);
		},
	);
});
